recipes-openxt/openxt/openxt-keymanagement/key-functions

```diff
@@ -0,0 +1,551 @@
+    This is the ${try} attempt.
+
+    Please enter the recovery password " 0 0 2>${capture} 1>&3
+
+    local ret=$?
+    if [ ${ret} != 0 ]; then
+        rm -f ${capture}
+        exec 3>&-
+        return ${ret}
+    fi
+
+    exec 3>&-
+    echo ${capture}
+}
+
+# Seed kernel's CPRNG with some entropy
+# This is a cheap attempt to get the entropy pool to an acceptable
+#   level. If there is a concern then override this function to call
+#   out to an external entropy source.
+seed_entropy() {
+    local seeks="5 72 96 128 51 83 112 12 31 205"
+
+    if [[ ! -e /dev/sda ]]; then
+        return 1
+    fi
+
+    for seek in $seeks; do
+        local entropy=$(cat /proc/sys/kernel/random/entropy_avail)
+        if [[ $entropy -gt 128 ]]; then
+            return 0;
+        fi
+
+        dd if=/dev/sda of=/dev/null bs=1M count=1 seek=${seek} 2>/dev/null
+        sync
+        sleep 2
+    done
+
+    if [[ $entropy -gt 128 ]]; then
+        return 0;
+    else
+        return 1;
+    fi
+}
+
+# Generates the TPM own key/password
+gen_own_key() {
+    local key_file=""
+    local well_known="0000000000000000000000000000000000000000"
+
+    if ! keystore_ready; then
+        setup_keystore
+
+        # If function 'answerfile_specifies' exist, then inside the installer
+        type answerfile_specifies 2>/dev/null | head -1 | grep -q "function$"
+        if [[ $? -eq 0 ]]; then
+            if interactive; then
+                show_cursor
+                key_file=$(password_prompt "TPM Own" "own")
+                hide_cursor
+            else
+                key_file=${key_file:-"${KEYDIR}/own"}
+                echo -n ${well_known} > ${key_file}
+            fi
+        else
+            key_file=$(password_prompt "TPM Own" "own")
+        fi
+    fi
+
+    key_file=${key_file:-"${KEYDIR}/own"}
+
+    if [ ! -e ${key_file} ]; then
+        echo -n ${well_known} > ${key_file}
+    fi
+
+    echo ${key_file}
+}
+
+# Generates the recovery key/password
+gen_recovery_key() {
+    local key_file=""
+    local well_known="0000000000000000000000000000000000000000"
+
+    if ! keystore_ready; then
+        setup_keystore
+
+        # If function 'answerfile_specifies' exist, then inside the installer
+        type answerfile_specifies 2>/dev/null | head -1 | grep -q "function$"
+        if [[ $? -eq 0 ]]; then
+            if interactive; then
+                show_cursor
+                key_file=$(password_prompt "device recovery" "recovery")
+                hide_cursor
+            else
+                key_file=${key_file:-"${KEYDIR}/recovery"}
+                echo -n ${well_known} > ${key_file}
+            fi
+        else
+            key_file=$(password_prompt "device recovery" "recovery")
+        fi
+    fi
+
+    key_file=${key_file:-"${KEYDIR}/recovery"}
+
+    if [ ! -e ${key_file} ]; then
+        echo -n ${well_known} > ${key_file}
+    fi
+
+    chcon -t lvm_tmp_t ${key_file}
```

Comment:
for chcon, because the previous is for the tpm owner key, not used by cryptsetup for lvm luks key. I would really like to get away from it because it's not a good practice, open to suggestions. for refactoring, this was already an abstraction from what was done in the existing approach to produce a clean key management api.